server/src/main/java/org/apache/lucene/index/LazySoftDeletesDirectoryReaderWrapper.java

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.apache.lucene.index;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import org.apache.lucene.document.Field;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.DocValuesFieldExistsQuery;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.FixedBitSet;
+import org.elasticsearch.common.lucene.Lucene;
+
+/**
+ * This is a modified version of {@link SoftDeletesDirectoryReaderWrapper} that materializes the liveDocs
+ * bitset lazily. In contrast to {@link SoftDeletesDirectoryReaderWrapper}, this wrapper can only be used
+ * for non-NRT readers.
+ *
+ * This reader filters out documents that have a doc values value in the given field and treat these
+ * documents as soft deleted. Hard deleted documents will also be filtered out in the live docs of this reader.
+ * @see IndexWriterConfig#setSoftDeletesField(String)
+ * @see IndexWriter#softUpdateDocument(Term, Iterable, Field...)
+ * @see SoftDeletesRetentionMergePolicy
+ */
+public final class LazySoftDeletesDirectoryReaderWrapper extends FilterDirectoryReader {
+    private final String field;
+    private final CacheHelper readerCacheHelper;
+    /**
+     * Creates a new soft deletes wrapper.
+     * @param in the incoming directory reader
+     * @param field the soft deletes field
+     */
+    public LazySoftDeletesDirectoryReaderWrapper(DirectoryReader in, String field) throws IOException {
+        this(in, new LazySoftDeletesSubReaderWrapper(Collections.emptyMap(), field));
+    }
+
+    private LazySoftDeletesDirectoryReaderWrapper(DirectoryReader in, LazySoftDeletesSubReaderWrapper wrapper) throws IOException {
+        super(in, wrapper);
+        this.field = wrapper.field;
+        readerCacheHelper = in.getReaderCacheHelper() == null ? null : new DelegatingCacheHelper(in.getReaderCacheHelper());
+    }
+
+    @Override
+    protected DirectoryReader doWrapDirectoryReader(DirectoryReader in) throws IOException {
+        Map<CacheKey, LeafReader> readerCache = new HashMap<>();
```

Comment:
This method should not be called in ES because this wrapper is used in ReadOnlyEngine only. The code might be simpler if we remove the "reader cache list". But I am fine if you keep it.

Reply:
+1 to throw an exception instead

Reply:
fixed in ba54eb3